### PR TITLE
fix: [CDS-79786]:  Update repoUrl and charName when path is part of connector URL

### PR DESCRIPTION
**Pull request by alebuh**
## Describe your changes
When OCI Helm Connector was configured with a URL that contains a path (`somehost.io/path1/path2`) and not just the hostname (`somehost.io`) fetching chart versions failed.
Introducing changes that if the path is provided as part of the connector URL, we should extract it and add it as the prefix to the basePath and chartName.

Screenshot 2023-09-26 at 10 52 24

Screenshot 2023-09-26 at 10 53 49

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/53352)

### CODE REVIEW

**Comment by alebuh** [comment]
trigger uts

**Comment by alebuh** [comment]
trigger uts, messagecheck

**Comment by alebuh** [comment]
trigger messagecheck

**Comment by alebuh** [comment]
trigger messagecheck

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/OciHelmApiHelperUtils.java [hunk 1]
```diff
@@ -76,4 +85,22 @@ private Retrofit getRetrofit(Converter.Factory converterFactory, String baseUrl)
...
+
+    if (isEmpty(path)) {
+      return chartNameNormalized;
+    }
+
+    if (isNotEmpty(uriObject.getHost())) {
+      return path + "/" + chartNameNormalized;
+    } else {
+      if (path.contains("/")) {
+        return path.substring(path.indexOf('/') + 1) + "/" + chartNameNormalized;
```
Comment (markoharness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/OciHelmApiHelperUtils.java:101]: Isn't this going to give back double slashforward chars? xxxxx//xxxxx?

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/OciHelmApiHelperUtils.java [hunk 2]
```diff
@@ -76,4 +85,22 @@ private Retrofit getRetrofit(Converter.Factory converterFactory, String baseUrl)
...
+  String formatChartNameWithUrlPath(String repoUrl, String chartNameNormalized) throws URISyntaxException {
+    URI uriObject = new URI(repoUrl);
+    String path = normalizeFieldData(uriObject.getPath());
+
+    if (isEmpty(path)) {
+      return chartNameNormalized;
+    }
+
+    if (isNotEmpty(uriObject.getHost())) {
+      return path + "/" + chartNameNormalized;
```
Comment (markoharness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/OciHelmApiHelperUtils.java:98]: Can you use String.format method for this?

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/OciHelmApiHelperUtils.java (hunk 1 above)
Comment (alebuh) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/OciHelmApiHelperUtils.java:101]: It shouldn't be supposed to. chartNameNormalized will never have '/' at a beginning as it's normalized and from path we are picking from the first char after the '/'
If you have any specific test case that is bothering you I can add it to the test so we can be sure.

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/OciHelmApiHelperUtils.java (hunk 2 above)
Comment (alebuh) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/OciHelmApiHelperUtils.java]: done

**Review by markoharness — APPROVED** [review]

**Comment by markoharness** [comment]
trigger review

**Review by abhinavsingh1196 — APPROVED** [review]

**Pull request merged**